controllers/addon_controller_test.go

```diff
@@ -27,114 +22,94 @@ var (
 
 const timeout = time.Second * 5
 
+var instance *v1alpha1.Addon
+var updated *v1alpha1.Addon
+var fetchedAddon *v1alpha1.Addon
+var addonController *Controller
+
 var _ = Describe("AddonController", func() {
 
 	Describe("Addon CR can be reconciled", func() {
-		var instance *v1alpha1.Addon
-		var wfv1 = &unstructured.Unstructured{}
-		wfv1.SetGroupVersionKind(schema.GroupVersionKind{
-			Kind:    "Workflow",
-			Group:   "argoproj.io",
-			Version: "v1alpha1",
-		})
-
 		It("instance should be parsable", func() {
-			addonYaml, err := ioutil.ReadFile("../docs/examples/clusterautoscaler.yaml")
+			addonYaml, err := ioutil.ReadFile("./tests/clusterautoscaler.yaml")
 			Expect(err).ToNot(HaveOccurred())
 
 			instance, err = parseAddonYaml(addonYaml)
 			Expect(err).ToNot(HaveOccurred())
 			Expect(instance).To(BeAssignableToTypeOf(&v1alpha1.Addon{}))
 			Expect(instance.GetName()).To(Equal(addonName))
+
 		})
 
 		It("instance should be reconciled", func() {
 			instance.SetNamespace(addonNamespace)
-			err := k8sClient.Create(context.TODO(), instance)
-			if apierrors.IsInvalid(err) {
-				Fail(fmt.Sprintf("failed to create object, got an invalid object error. %v", err))
-			}
-			Expect(err).NotTo(HaveOccurred())
+			addonController = newController(instance)
 
-			Eventually(func() error {
-				if err := k8sClient.Get(context.TODO(), addonKey, instance); err != nil {
-					return err
-				}
-
-				if len(instance.ObjectMeta.Finalizers) > 0 {
-					return nil
-				}
-				return fmt.Errorf("addon is not valid")
-			}, timeout).Should(Succeed())
+			processed := addonController.processNextItem(ctx)
+			Expect(processed).To(BeTrue())
+			fetchedAddon, err := addonController.addoncli.AddonmgrV1alpha1().Addons(testNamespace).Get(ctx, testAddon, metav1.GetOptions{})
+			Expect(err).To(BeNil())
+			Expect(fetchedAddon).NotTo(BeNil())
 
 			By("Verify addon has been reconciled by checking for checksum status")
-			Expect(instance.Status.Checksum).ShouldNot(BeEmpty())
+			Expect(fetchedAddon.Status.Checksum).ShouldNot(BeEmpty())
 
 			By("Verify addon has finalizers added which means it's valid")
-			Expect(instance.ObjectMeta.Finalizers).Should(Equal([]string{"delete.addonmgr.keikoproj.io"}))
-
+			Expect(fetchedAddon.ObjectMeta.Finalizers).Should(Equal([]string{"delete.addonmgr.keikoproj.io"}))
 			oldCheckSum := instance.Status.Checksum
-
 			//Update instance params for checksum validation
 			instance.Spec.Params.Context.ClusterRegion = "us-east-2"
-			err = k8sClient.Update(context.TODO(), instance)
-
-			// This sleep is introduced as addon status is updated after multiple requeues - Ideally it should be 2 sec.
-			time.Sleep(5 * time.Second)
+			err = addonController.handleAddonUpdate(ctx, fetchedAddon)
```

Comment:
We shouldn't have to set the controller like this for this test, I think we said you need to setup the controller in the beforeSuite.

Reply:
2d17325c0d5fc3cf05adad48f62c54eccbb786e3